staging/src/k8s.io/kube-aggregator/pkg/controllers/openapi/aggregator/downloader.go

```diff
@@ -21,31 +21,54 @@ import (
 	"fmt"
 	"net/http"
 	"strings"
+	"sync"
 
 	"k8s.io/apiserver/pkg/authentication/user"
 	"k8s.io/apiserver/pkg/endpoints/request"
-	"k8s.io/kube-openapi/pkg/cached"
 	"k8s.io/kube-openapi/pkg/validation/spec"
 )
 
+type CacheableDownloader interface {
+	UpdateHandler(http.Handler)
+	Get() (*spec.Swagger, string, error)
+}
+
 // cacheableDownloader is a downloader that will always return the data
 // and the etag.
 type cacheableDownloader struct {
+	name       string
 	downloader *Downloader
 	handler    http.Handler
 	etag       string
 	spec       *spec.Swagger
+	mutex      sync.RWMutex
 }
 
 // NewCacheableDownloader creates a downloader that also returns the etag, making it useful to use as a cached dependency.
-func NewCacheableDownloader(downloader *Downloader, handler http.Handler) cached.Value[*spec.Swagger] {
+func NewCacheableDownloader(apiServiceName string, downloader *Downloader, handler http.Handler) CacheableDownloader {
 	return &cacheableDownloader{
+		name:       apiServiceName,
 		downloader: downloader,
 		handler:    handler,
 	}
 }
+func (d *cacheableDownloader) UpdateHandler(handler http.Handler) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+	d.handler = handler
+}
 
 func (d *cacheableDownloader) Get() (*spec.Swagger, string, error) {
+	d.mutex.RLock()
+	defer d.mutex.RUnlock()
+	spec, etag, err := d.getLocked()
```

Comment:
it's a little confusing to call `getLocked` when we're only under an RLock ... it's not obvious what mutex is guarding, but we assign to etag and spec inside `getLocked`

Reply:
I agree that's wrong. I don't know if it'll trigger with the current way things are implemented, but that's still a problem with that object.

Reply:
Yeah I agree. I was originally thinking of only protecting the handler with this lock, but I can see how this can be interpreted as etag/spec as well.

Reply:
I've changed the mutex to just a regular mutex since get is only called from one thread and rlock has no performance benefit. Also modified the lock to only be active when fetching the handler rather than the entire get operation.